### ISSUE TITLE
import of FIT 2.0 files in pre-release 3.4 results in weird data

### ISSUE THREAD

**Issue by gernot-hochmuth** [GoldenCheetah/GoldenCheetah#2072]
Recording SmO2 and tHb data using a Moxy sensor with a Garmin Vivoactive watch using a ConnectIQ 2.0 data field provided by the manufacturer of Moxy sensors in the standard Garmin Running app results in a FIT 2.0 file that cannot be imported properly in the pre-release 3.4 (DEV1608) of GoldenCheetah. I am not quite sure whether this is solely an issue with the pre-release of GoldenCheetah or whether there is some issue with this specific data field or with the implementation of ConnectIQ Developer (or Contributor as Garmin calls it) fields. I cannot verify this as the recorded Moxy data does not show up on Garmin connect nor in the Garmin Connect Android App. The  watch is updated to the most recent firmware which is suposed to allow recording of FIT 2.0 (ConnectIQ Contributor) data.
Do other users of Garmin Devices experience similar issues with ConnectIQ Contributor data on Garmin Connect? Do other users experience similar issues when importing FIT 2.0 data into GC 3.4 pre-release?

The following anomalies appear in the respective graphs and tables in GoldenCheetah:

- Heart rate data is apparently overwritten by the value of tHb times 100, giving values around 1250 in my case! 
- Altitude is off by approximately -350 m (I live at an altitude of ca. 100 m, the "new" values vary around -250 m).
- Latitude and Longitude are apparently nonsense values (Latitude ranging from 2 to 4 deg, longitude stays close to 0 - that is somewhere in the atlantic ocean south of Ghana and west of Sao Tome, while running in Germany). I have no idea which other values could have overwritten latitude and longitude.

Moxy data is not shown in the graph of the OXY section in GoldenCheetah (for some strange reason it's temperature data that is shown in the OXY section - but that is not specific to the pre-release 3.4). Moxy data can be tabulated in the DEVELOPER tab of the EDIT section in GoldenCheetah. In the DEVELOPER tab the following data is listed:

- first column: time (plausible values)
- second column: distance (all values set to 0, but plausible distance data is listed in the STANDARD tab)
- third column: SmO2-sensor (apparently SmO2 times 10, e.g. 390 instead of 39 %)
- fourth column: tHb-sensor (apparently tHb times 100, e.g. 1250 instead of 12.50 g/dl), these values are apparently assigned to the heart rate data (see above)
- fifth column: SmO2-sensor (plausible values, e.g. 39 %). same values as third column divided by 10
- sixth column: tHb-sensor (plausible values, e.g. 12.50 g/dl). same values as fourth column divided by 100

The column headers of the DEVELOPER tab are given in english whereas the column headers of the STANDARD tab are given in local language (german in my case) in the EDIT section.
Moxy data is not summarized in the SUMMARY section (as it is when you import separately recorded Moxy data).
Should I worry about 12 billion TRIMP points? This used to be much lower beforehand.

**Comment by grauser** [comment]
Dear gernot-hochmuth,

Could you send your file (or better files) to me ?

Thank you.

Damien

Le dim. 28 août 2016 à 22:58, gernot-hochmuth <notifications@github.com> a
écrit :

**Comment by gernot-hochmuth** [comment]
Hi Damien,

sure. You will find some files in the ZIP-folder on Dropbox:

https://www.dropbox.com/s/4c5quqff7px5io3/garmin_fit_2_0.zip?dl=0


In the ZIP I have put three FIT files and three corresponding CSV-files.
I managed to convert the FIT files to CSV using the FITtoCSV.jar tool
from the recent FIT-SDK 20.10. Developer data (SmO2 and tHb) are
included in the csv-files. SmO2 and tHb do not show up on Garmin Connect.
I have recorded the data with my Vivoactive watch with the most recent
firmware 4.20 using Garmin's proper running app. To record the data I
had to use either the Moxy-datafield provided by Moxy on the ConnectIQ
app store (file 68S82601.FIT in the ZIP) or a self compiled
Moxy-datafield example from Garmin's ConnectIQ-SDK (files 692K2620.FIT
and 1336960917.FIT in the ZIP).
The precompiled version 3.4 of Goldencheetah for windows does not show
"Developer" data in the edit tab. A self compiled version 3.4 of
GoldenCheetah for Linux (including some contributions from GIT) shows
"Developer" data (i.e. SmO2 and tHb) in the edit tab. In both versions
(Windows or Linux) I find strange data for heart-rate, altitude,
latitude and longitude as reported before.

Heart-rate seems to be overwritten by tHb * 100 when using the datafield
provided by Moxy or by SmO2 * 10 when using the example datafield from
the ConnectIQ SDK.

Altitude gives nonsense values when using the datafield provided by Moxy
but seems to be plausible (though not very precise) when using the
example datafield from the ConnectIQ SDK (this data may represent
something else than altitude as well).

latitude and longitude give nonsense values in any case.

Other than stated before I find different data tabulated in the
developer tab of the Edit section when using the ConnectIQ example
datafield:

first column: time (Moxy and ConnectIQ sample)
second column: distance = 0 (Moxy and ConnectIQ sample)
third column: SmO2 * 10 (Moxy) or tHb * 100 (ConnectIQ sample)
fourth column: tHb * 100 (Moxy) or lap-tHb = 0 (ConnectIQ sample)
fifth column: SmO2 (Moxy) or average-tHb = 0 (ConnectIQ sample)
sixth column: tHb (Moxy) or SmO2 * 10 (ConnectIQ sample)
seventh column: lap-SmO2 = 0 (ConnectIQ sample)
eighth column: average-SmO2 = 0 (ConnectIQ sample)

Looking at the data section of the CSV-files I notice some rows with
differing data-order. These rows are referred to as "laps", "events" and
"unknown".

I hope that helps a little.
I will remove the ZIP from Dropbox within the next two weeks.

regards

Gernot


Am 02.09.2016 um 12:07 schrieb Damien Grauser:
